### DISCUSSION TITLE
Let's start the Transmission 4.1 development?

### DISCUSSION THREAD

**Discussion by Coeur** [transmission/transmission#4736] [Ideas]
@ckerr @mikedld Congratz on Transmission 4.0.

Since we have a few pending PRs that are adding strings needing to be translated, I suggest that we focus on Transmission 4.1 development to deal with those?

And once they are tackled, we can consider working on Transmission 5 for larger changes.
What do you think?

**Comment by svin24** [comment]
well there is this bug that really hurts adoption in platforms like windows https://github.com/transmission/transmission/issues/255 might want to focus on that

**Comment by ckerr** [comment]
I like the idea of unblocking the [decide-after-4.0.0](https://github.com/transmission/transmission/milestones/decide-after-4.0.0) items but to me the higher short-term priority is to make sure that 4.0.x is stable. I'd like to see how the bug reports for 4.0.0 go before landing [non-patch changes](https://semver.org/#spec-item-6).

**Reply by Coeur** [reply]
Can we have multiple branches then, instead of many pending PRs?
One branch for 4.0, one for 4.1 (which possibly is the main branch).
Any change made on branch 4.0 would then be forwarded to 4.1 (which can maybe be done with GitHub actions).

**Reply by ckerr** [reply]
Yes, of course branching is what will happen. I am a little wary of branching "too soon", e.g. I'd like to have a feel for how many bugs get reported on 4.0.0 because every bugfix will have to go into both branches and can be more complicated e.g. if the code around the fix changes during the 4.1.x work.

I guess I'd ask which PRs you specifically have in mind? Might be easier to take it on a case-by-case basis.

**Reply by ckerr** [reply]
To clarify the prev ccomment -- we are getting a lot of bug reports right now. No _blockers_ per se -- e.g. no data corruption or the like -- but still, a lot of people are testing out 4.0.0 and finding things. I'd _really_ like it if all of us would focus on those right now instead of landing new features.

There have been about 20 commits since 4.0.0. If we'd branched right after 4.0.0 then each one of those would have to be merged twice, once to `main` and once to `4-0-x`. 

> Any change made on branch 4.0 would then be forwarded to 4.1 (which can maybe be done with GitHub actions).

Who is volunteering to write the Action to do that? :smile_cat:

**Reply by Coeur** [reply]
@fetzu would you know how to do that GitHub action?

**Reply by fetzu** [reply]
If I understand the posts above correctly, there would surely be a way to do this through Actions.

Just to make sure I understand the workflow you are aiming for: you are looking to use branches (e.g., '4-0-1' for the next patch release; '4-1-0' for the next minor release, and '5-0-0' for the next major release) concurrently for development but want to make sure that changes to 4-0-1 are also mirrored to 4-1-0 and 5-0-0 automatically ? Are you still expecting for the commits/PRs to be done into 'main' as they (mostly) are now?

One way which might be doable is to use tags in the PRs: maintainer sets the tags the patch applies to before merging, and the Action automatically creates/merges a corresponding PR in all the branches (for a PR tagged with '4-0-1' to '4-1-0' and '5-0-0', for a PR tagged with '4-1-0' only to '5-0-0'...). Assuming there are no conflicts, this _could_ work.

**Reply by Coeur** [reply]
Just two branches: main and 4.1, and whatever is merged on main is forwarded to 4.1.
You can ignore 5.0 for now: what we want to do is tackle opened PRs (by merging them on 4.1) instead of keeping them opened for too long, often getting in conflict with passing time.

**Reply by fetzu** [reply]
Got it. Just a couple of quick follow-ups: 
1. How should the 4.1 branch be called? dev/4-1-0 ?
2. What should the trigger be push to the 4.1 branch? Each push on main or just merged PRs?

**Reply by Coeur** [reply]
I think we can use dots in branch names, so better 4.1 than 4-1.
I'll let the owners decide (ckerr and/or mikedld)

**Reply by fetzu** [reply]
OK, I have a basic action working (that just pushes any push from main into a dev branch). The commits done on main will be replicated 1:1 to the dev branch.

I just need the maintainers to weigh in:
1. How do you want the development branch to be called (4.1, 4.1.0, dev/4.1, 4-1, 4-1-0, .... you get the idea ;)) ?

PS: I will also need the people with the right permissions to create the branch beforehand (I'll add that to the PR).

**Reply by Coeur** [reply]
Actually, never mind that, fetzu. I've changed my mind.
I now think that Transmission 4.0.1 will be quick to come, as seen by the speed at which each issue is being addressed. So my feeling is that I just have to wait a few days for 4.0.1 to be released and 4.1 development to be most active.

**Reply by ckerr** [reply]
Bold of you to assume there won't be more bug reports and we repeat the cycle for 4.0.2 :smile_cat: 

Even if that does happen IMO it's a Good Thing because we _want_ a stable 4.0.x before moving on to new features

**Reply by fetzu** [reply]
Well, if are interested of setting something up to help with your workflow (from a basic branch copying action to a more complex workflow with labels on PRs), I'd gladly do my best to lend a hand!

**Reply by ckerr** [reply]
@fetzu thanks!

Sure, what I had in mind for backporting patches was [trop](https://github.com/electron/trop) or something troplike.

Given the torrent of bug reports we've gotten over the last week, I'm convinced we (a) should release more often and (b) probably shouldn't be thinking about landing new features in the short term :smile_cat:  So you if you want to do this you have a little time to play around

**Comment by Pentaphon** [comment]
> Since we have a few pending PRs that are adding strings needing to be translated, I suggest that we focus on Transmission 4.1 development to deal with those?

I think 4.0.x needs work first. There's a lot of bugs that need to be fixed before the project can move on to 4.1.0. I wouldn't be surprised if 4.0.1 has to be pushed out by the end of Winter. 

We may end up with 2 or even 3 patch versions before the 4.1.x version can be worked on. 

In general, I don't think it's a good idea to move though the 4.x.x series quickly. After the rebase, I expect the project to spend a long time making the client stable on all 3 platforms. If the project added some auto-crash-logging, I suspect we'd see a lot of issues that need to be resolved but don't get reported by regular users.

I don't see any reason to go to 5.0.0 anytime in this decade. It's taken nearly 3 years to go from 3.0 to 4.0 with no versions in between.

**Reply by ckerr** [reply]
> I think 4.0.x needs work first. 

Agreed.

>  I wouldn't be surprised if 4.0.1 has to be pushed out by the end of Winter.

I'd quibble on the phrasing of "has to" since it kind of sounds like it would be released grudgingly due to a showstopper bug. There aren't any known blocker bugs (so far) and I'd be happy to do several patch releases as fixes are made.

> I suspect we'd see a lot of issues that need to be resolved.

Pull Requests Welcomed! :smile_cat: 

> I don't see any reason to go to 5.0.0 anytime in this decade.

If we're following semver we _do_ need to reconsider what Transmission 5 means.

The project has a lot of big-ticket features milestoned for 5 but in semver terms they are actually semver-minor (backwards-compatible new functionality) so things like BitTorrent v2 support and async IO could go into a 4-x-y release.

Ironically enough the major number would probably only need to change for non-user-facing changes, e.g. changes to RPC like https://github.com/transmission/transmission/issues/2855.

**Reply by Pentaphon** [reply]
> I'd quibble on the phrasing of "has to" since it kind of sounds like it would be released grudgingly due to a showstopper bug. There aren't any known blocker bugs (so far) and I'd be happy to do several patch releases as fixes are made.

Right now, I think the project is facing LOT of smaller bugs that when taken as a whole, will probably annoy new users like magnets starting as paused, the multiple instances issue, the port checker issue, the issue where the process something keeps running on Windows after being exited, etc. That's why I think a bunch of patch versions will be pushed out this year before the devs can move on to 4.1.x

> The project has a lot of big-ticket features milestoned for 5 but in semver terms they are actually semver-minor (backwards-compatible new functionality) so things like BitTorrent v2 support and async IO could go into a 4-x-y release.
> 
> Ironically enough the major number would probably only need to change for non-user-facing changes, e.g. changes to RPC like #2855.

Considering that the project went from 3.0 to 4.0.0 with no minor versions in between, I think everybody would and should expect this project to stay on 4.x.x for a very long time, but with somewhat more frequent patch releases similar to the 2.x era.

**Comment by Coeur** [comment]
*bump*

We could almost... almost... just merge all the pending 4.1.0 pull requests and make a 4.1.0 release within **1 day**. :D
OK, I'm exaggerating, but we likely won't need to have a too long development period for 4.1.0. The longest part should just be about getting the translations for new strings, in my opinion. Then we could do later a 4.2.0 for stuff that didn't make it in the first batch. Then a 5.0.0 for more drastic changes. Something like that.

**Reply by GaryElshaw** [reply]
Yep, i think so too. There is a lot of semver minor changes in the pipeline that sat on the cusp of being included in the current releases. And i don't think it's necessarily indicative of issues with the clients, but there are fewer bug reports since 0.2.

My 2c ideal next step or path forward? Less focus on piecemeal fixes, but time spent looking at performance parity between clients; where the codebase might be after so many iterative fixes and changes (preferably two maintainers or peer reviewed); and a build and publish for testing in an entirely new way.

1. Status quo with a focus on investigating performance parity between clients for a 4.1.0 release
2. Simultaneously, move to triage/merge semver minor into 4.1.0, immediately
3. Introduce a 4.5.0 experimental branch that is purely focused on everything future, and i mean a proper skunk works branch that allows creativity, debates and arguments about the entire foundations of the future of the clients.

**Reply by ckerr** [reply]
> but there are fewer bug reports since 0.2.

It's been released for like a day :smile_cat:

**Reply by GaryElshaw** [reply]
Weirdly, I am aware of this (i'm here daily and available for weddings), but i also remember the deluge of 4.0 and 4.0.1. 🫢

**Reply by GaryElshaw** [reply]
But the substance? I'd really like the project to take a step back to code review. And yes, i apologise, i know i'm not much help on that front.

**Reply by ckerr** [reply]
> But the substance?

> focus on investigating performance parity between clients for a 4.1.0 release

We are all volunteers. If someone steps up to do that work, wonderful. I am *always* interested in PRs that improve performance.

If there are actionable bug reports, that's a big help; e.g. https://github.com/transmission/transmission/issues/5206 is a performance-related issue that is a 4.1.0 priority largely because tearfur did the research and found actionable info.

Getting #4900 into 4.1.0 will probably help future performance issue triage.

> 4.1.0, immediately

Patches welcomed. https://github.com/transmission/transmission/issues/4614, https://github.com/transmission/transmission/issues/4790 are blockers

> Introduce an experimental branch

I'm not 100% against this but I'm not sold on it either. How is this preferable to a set of PRs for different ideas / topics?

>  I'd really like the project to take a step back to code review.

This criticism doesn't feel valid to me. In the last couple of days -- which have been slow -- @Coeur reviewed https://github.com/transmission/transmission/pull/5233, I reviewed https://github.com/transmission/transmission/pull/5244, Mike and I both reviewed https://github.com/transmission/transmission/pull/5249.

**Reply by GaryElshaw** [reply]
I'm really confused, Charles. It's not the first time that you and i misunderstand each other, but i'm always a bit surprised when these moments occur, where you perceive such differences, and i feel like i'm in some kind of weird parallel universe where i start feeling like an incompetent impramatur. 

My sugestion was that now is the time, i think, for a code review. Not by me, but by you and others to ascertain the current status of the code after a plethora of changes putting out fires between 4.0 and now.

I currently spend equal time in Linux and macOS, i think it's counter-intuitive to explore individual pr's and say that's code review. My point was to up and down explore the code in the same way that the C to C++ transition was done.

**Reply by ckerr** [reply]
Ultimately what matters is who is going to step up to do the work. Coding, translating, updating documentation, doing code review, answering questions, and reporting actionable bugs. Those are the things that get releases out the door.

So the question that matters is, who is going to do the research to come up with actionable info on performance parity? Who is going to do the project-wide code inspection? Who has these skunkworks PRs?

**Reply by ckerr** [reply]
> it's counter-intuitive to explore individual pr's and say that's code review.

That is literally the common definition of code review. 😸  https://about.gitlab.com/topics/version-control/what-is-code-review/

>  My point was to up and down explore the code in the same way that the C to C++ transition was done.

OK. I would usually call that a walk-through or a code inspection.

**Reply by GaryElshaw** [reply]
I don't perceive these things as being lumped in together, it's why i suggested that 0.2 out the door might now be a time to take stock. A lull, to step back for you and others contributing code to examine the codebase, with less thought focused on immediacy, but where things are. I'm actually less concerned with rushing anything out the door than i am in where the code might be after so many iterative changes that have often been putting out fires.

Charles, I am fully aware that you lead this project, and i am also aware that there is an onus that falls on you. But i think it might be time to also understand the others also have a vested interest and might possibly be credited with being Maintainers also. And no, this isn't about me. Coeur should become a maintainer. To be honest, he is already, and i think he deserves that credit to the project whether he wants it or not. 

You know my day job is web, accessibility, and getting things out the door, yes? I'm sugesting now is the optimal time to code review. The fires are out, so what seems important next? It's not my project, i'm only a contributor, but i do manage projects IRL. I'm suggesting taking a breath.

**Reply by ckerr** [reply]
> and i think he deserves that credit to the project whether he wants it or not...

1. Did you ask Coeur what they want, or no?

2. Attribution & credit is a valid topic that deserves its own thread. E.g. the credits should be cleaned up to give proper respect. I haven't thought about write permissions and it's worth thinking about. Right now IIRC five people have write permissions.

> investigating performance parity between clients for a 4.1.0 release ... Simultaneously, move to triage/merge semver minor into 4.1.0, immediately ... Introduce a 4.5.0 experimental branch that is purely focused on everything future ...  now is the time, i think, for a code review. Not by me, but by you and others ...

>  I suggested [a] lull ... I'm suggesting taking a breath.

*that's* what you call a lull?? 🤣 💫 

Anyway. Jokes aside... proposals are great but the question that matters is, who is going to step up and do the work, and what do they want to work on.

For example, if client performance disparity is the issue you want to champion, but you don't want to code... go do the research that figures out if this is actually a problem; and if so, **why it happens**, e.g. does the Windows build spend more time in disk IO than the Linux build, or whatever -- I.e. an actionable bug report. Doing the work is what adds value.

**Reply by GaryElshaw** [reply]
> but you don't want to code...

Outside of web, I **can't** code


> go do the research

I did do some. https://github.com/transmission/transmission/issues/5003#issuecomment-1446332609 and https://github.com/transmission/transmission/issues/5003#issuecomment-1449278351

Leading me to think it has something to do with how the Linux clients are engaging with peers, or there is something going on potentially with how Transmission is interacting with the linux network stack.

**Reply by Coeur** [reply]
off topic reply: I do not wish to become a Maintainer or be credited for present and past contributions. In the future I may change my mind for some contributions (like the Swift conversion), but if I'm not explicit about it, then please consider that I still prefer to stay low profile. I'm aware that my email address is in the commits and that's enough publicity. I never even asked what is "Mnemosyne LLC" or who are the receivers of [donations](https://transmissionbt.com/donate.html), but I don't want to receive any money. Sorry for bringing that.

Back to topic: I trust Charles and Mike in handling the release schedules and deliveries. And I hope that one day we'll be on par with qBittorrent download performances (which isn't the case yet, sadly, at least on macOS).

**Reply by ckerr** [reply]
> > go do the research
> 
> I did do some. [#5003 (comment)](https://github.com/transmission/transmission/issues/5003#issuecomment-1446332609) and [#5003 (comment)](https://github.com/transmission/transmission/issues/5003#issuecomment-1449278351)
> 
> Leading me to think it has something to do with how the Linux clients are engaging with peers, or there is something going on potentially with how Transmission is interacting with the linux network stack.

Cool. That's a great start; run with that! :tada: 

Right now there's nothing actionable so I don't know where to take that ticket, but I want you to understand that the takeaway here is to be encouraged to do more research. I'm not shooting you down; I'm just saying that "something potentially with the linux network stack" is not actionable so more research is needed :green_heart:

**Reply by ckerr** [reply]
> I never even asked what is "Mnemosyne LLC" or who are the receivers of [donations](https://transmissionbt.com/donate.html), but I don't want to receive any money. Sorry for bringing that.

Mnemosyne is an LLC that I formed at one point back in the day when reading the judgements in the Napster & Grokster lawsuits. We're all volunteers and don't have big corporate money to pay for lawyers, so I figured it would be good to have a legal entity there to act as a firewall.

As for money... It was reeeeally dry before 4.0.0. There has been a 4.0.x bump. Not "donos bought my car" money; more like "donos bought my car's gas" money + "nice dinner with the wife" money. If the release cadence is going to be higher, maybe the donos will continue and we should think about this. I'm not sure how to do this fairly and the potential for hurt feelings is really high. **So for anyone who is doing work and wants a cut, including but not just @Coeur, please reach out to me and lmk.** I want us to be a happy family and so any imbalance is more about this being a potential minefield of hurt feelings and of me not being sure who wants what and how to do this without stepping on a mine.

**Reply by Coeur** [reply]
Hi. Minor misunderstanding: I do not want money and I can't accept money. But I can donate. :)

**Reply by Coeur** [reply]
By the way, **I may be away from internet between March 22-April 5**.

**Comment by ckerr** [comment]
Probably yes. I'd like to wait and see if the release of `4.0.2` generates another round of bug reports, but... the number of non-duplicate bugs  being submitted has gone back down to a sane level lately, so it does feel like we're getting closer to letting `semver/minor` code into `main` again.

**Reply by ckerr** [reply]
If the # of bugs reported on 4.0.2 stays pretty low then IMO we should open up to 4.1.0 commits sometime next week

**Comment by vsheg** [comment]
@ckerr the release of 4.1 with sequential download support is highly anticipated

**Comment by ckerr** [comment]
@Coeur, @GaryElshaw if y'all are in a hurry to move to `4.1.0` please self-assign #4614, #4790 and fix them, that seems like a high priority. :bow: 

Those need to be fixed in a `4.0.x` release _before_ `4.1.0-beta.1` is released so we don't screw up updates in another beta cycle...

**Reply by GaryElshaw** [reply]
I could be totally wrong, but 4614 and 4790 required a maintainer to address that, i think? @Coeur might remember, because i think that required information or login information that maybe @livings124 exclusively posessed?

**Reply by GaryElshaw** [reply]
By-the-by, i'm all about not screwing up macOS updates, but i'm not sure 4.1.0 needs a beta status. How about we do something i suggested a year ago? We better communicate the project status and have more folks willing to test?

We're partially there with recent changes to directing folks to the CI builds; i'd like my suggestion to be debated with the maintainers that a separate set of builds are sent to Jenkins that make the process easier for possible testers, with clear caveats.

**Reply by Coeur** [reply]
I tried to address it, but it didn't pass code review. I don't think we should continue to block 4.1.0 development because of that issue, since hardly no one but the maintainers themselves can deal with it, as Sparkle is a "release" tool.

**Comment by ckerr** [comment]
Feedback on 4.0.2 seems pretty good so far, so [as mentioned last week](https://github.com/transmission/transmission/discussions/4736#discussioncomment-5329198) maybe it's time to make a stability branch for 4-0-x and start landing `semver/minor` changes.

Apparently everyone's been waiting for this, but just to make sure -- if anyone is against this idea and thinks we should stay in bugfix-only mode longer, please give a shout and say why!

**Reply by Coeur** [reply]
If there are only 3 pull requests pending for 4.0.3, we can merge them and release 4.0.3 before starting 4.1.0 next week.

**Reply by ckerr** [reply]
I'm traveling next week and probably won't be available for a release. 🤷‍♂️

**Reply by Pentaphon** [reply]
> If there are only 3 pull requests pending for 4.0.3, we can merge them and release 4.0.3 before starting 4.1.0 next week.

It's too early for 4.0.3. I think you guys need to give those bug reports some time to come in and then do a 4.0.3 at the end of April. 4.1.0 is going to take a long time so you can easily give 4.0.3 another month or so.

**Reply by Coeur** [reply]
4.0.3 is fixing a crash for non-utf8 magnets: I don't think that we should wait one month to release that crash fix.

**Reply by Coeur** [reply]
4.0.3 was released. Let's start 4.1.0 one way or another. Either main branch or separate branch, but let's start merging 4.1.0 stuff. Thanks.

**Comment by ckerr** [comment]
4.1.0-dev is in development now.
